packages/video_player/video_player_android/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayer.java

```diff
@@ -40,6 +42,7 @@
 import java.util.List;
 import java.util.Map;
 
+@OptIn(markerClass = UnstableApi.class)
```

Comment:
updated, is it okay?

Reply:
so, not migration to media3 then (at least until there is a stable alternative for such apis?)
or am I missing something?

Reply:
another solution I see is adding non-endorsed plugin - video_player_android_media3 until media3 stabilizes

Reply:
though, wasn't it the same all the time?
https://github.com/androidx/media/issues/331#issuecomment-1512711590

Reply:
The fact that the old code was silently doing something dangerous doesn't mean we should explicitly opt the entire plugin into continuing to do that, or putting a comment that describes it as safe when it isn't.

What is the exact list of APIs we are using that are unstable? We should be using highly targeted annotations, and we should investigate each of them to determine if they are truly necessary, and if so comment why, and about the danger they pose.

Reply:
>What is the exact list of APIs we are using that are unstable?

updated pr with fine grained opt-in
didn't find stable alternative for e.g. setMediaSource (whereas MediaSource itself is stable)

Reply:
do I understand it correctly - even if we are using stable only API - any client can still use unstable api on its own or with some dependency, which potentially breaks it

so
1. all good with whatever api if we are the only dependency
2. high chance things would break if both video_player_android and some other dependency use unstable api with different minor/patch versions
3. even we use stable only api - any other dependency with unstable api gives problems - if it depends on older version

however, I don't understand how even stable api helps if media3 bumps major version (legally changing stable api) - gradle would still prefer it by default afaik

Reply:
> any client can still use unstable api on its own or with some dependency, which potentially breaks it

I don't follow. How would plugin code using only stable APIs be broken by something else using unstable APIs?

> even we use stable only api - any other dependency with unstable api gives problems - if it depends on older version

If some other plugin chooses to use unstable APIs, and doesn't do any strict version control, then *it* could break, yes. That's the nature of using unstable APIs. We can't control what other developers choose to do in their own plugins. We can only attempt to make our own code robust and stable.

> however, I don't understand how even stable api helps if media3 bumps major version (legally changing stable api) - gradle would still prefer it by default afaik

Gradle does optimistic upgrades across *major* version changes? If so, that's bad, and we should revisit our versioning constraints in general. /cc @reidbaker Are you familiar with the details of how optimistic upgrades work?

Reply:
> I don't follow. How would plugin code using only stable APIs be broken by something else using unstable APIs?

Plugin won't break its code - it would break client code if one uses older version
e.g. client uses setMediaSource in version 1.1.1, which is removed in version 1.4.0 which plugin uses (despite its all stable api in plugin)

Reply:
> Gradle will consider all requested versions, wherever they appear in the dependency graph. Out of these versions, it will select the highest one. More information on version ordering [here](https://docs.gradle.org/current/userguide/single_versions.html#version_ordering).

https://docs.gradle.org/current/userguide/dependency_resolution.html

> In case of conflict, Gradle by default uses the newest of conflicting versions

https://docs.gradle.org/current/dsl/org.gradle.api.artifacts.ResolutionStrategy.html#org.gradle.api.artifacts.ResolutionStrategy:dependencySubstitution

Reply:
@stuartmorgan I think this comment is saying that gradle considers all dependencies in the scope of the app which means that versions specified in the plugin could be overridden by dependencies the app (or another plugin) has. Which both feels like a gotcha that might exist and also not how build systems should work and is not something I was aware of before.

Reply:
I think it's much better to clearly scope the use of unstable APIs as the current commit does, if (as sounds like is the case) we are stuck with using them for now. I would also comment the annotations with a TODO pointing to an issue discussing the issue and the need to track a long term solution.

Based on the discussion above it sounds like:
- Short term, we are no worse off than before by moving forward with the migration (but @reidbaker and @camsim99 should weigh in as well) in terms of potential for the plugin to stop compiling for clients due to upstream changes.
- Long term, we need to revisit how we express dependencies in our Android plugins, to see if we can prevent optimistic updates across major versions. I'll file an issue for that.
- Long term, we need to revisit the use of unstable APIs in this plugin (the TODO).

Reply:
added TODOs and created https://github.com/flutter/flutter/issues/147039